test/skimNtuple2016_HHbtag.cpp

```diff
@@ -974,56 +896,82 @@ int main (int argc, char** argv)
     float stitchWeight = 1.0;
     if (DY_tostitch)
     {
-      int njets = theBigTree.lheNOutPartons;
-      int nb    = theBigTree.lheNOutB;
-      // these protections should be useless
-      if (njets < 0) njets = 0;
-      if (njets > 4) njets = 4;
-      if (nb < 0)    nb = 0;
-      if (nb > 4)    nb = 4;
-
-      float ht = theBigTree.lheHt;
-      int nht = 0;
-      if      (ht  < 0                ) nht = 0;
-      else if (ht >= 0    && ht < 70  ) nht = 0;
-      else if (ht >= 70   && ht < 100 ) nht = 1;
-      else if (ht >= 100  && ht < 200 ) nht = 2;
-      else if (ht >= 200  && ht < 400 ) nht = 3;
-      else if (ht >= 400  && ht < 600 ) nht = 4;
-      else if (ht >= 600  && ht < 800 ) nht = 5;
-      else if (ht >= 800  && ht < 1200) nht = 6;
-      else if (ht >= 1200 && ht < 2500) nht = 7;
-      else  /* ht >= 2500 */            nht = 8;
-
-      stitchWeight = stitchWeights[njets][nb][nht];
-    }
-
-    // Should never enter here (DY_tostitch should be always true)
-    if (!DY_tostitch && DY_nJets >= 0)
-    {
-      int njets = theBigTree.lheNOutPartons;
-      int nb    = theBigTree.lheNOutB;
-      //cout << "- njets: " << njets << " - nb: " << nb << endl; //FRA debug
-      if (njets != DY_nJets || nb != DY_nBJets) continue;
-    }
-
-    if (DEBUG && isMC)
-    {
-      cout << "** DEBUG : gen particle list" << endl;
-      for (unsigned int igen = 0; igen < theBigTree.genpart_pdg->size(); igen++)
+      float Zpt = -999;
+      // loop through gen parts to identify Z boson
+      int Zidx = -1;
+      int idx_tau1 = -1;
+      int idx_tau2 = -1;
+      int pdg_tau1 = -1;
+      for (unsigned int igen = 0; igen < theBigTree.genpart_px->size(); igen++)
       {
-	int pdg = theBigTree.genpart_pdg->at(igen);
-	if (abs(pdg) == 66615 || abs(pdg) == 11 || abs(pdg) == 13)
+	bool isZLast   = CheckBit(theBigTree.genpart_flags->at(igen), 13) ; // 13 = isLastCopy
+	bool isZPrompt = CheckBit(theBigTree.genpart_flags->at(igen),  0) ; //  0 = isPrompt
+	bool isHardProcess = CheckBit(theBigTree.genpart_flags->at(igen),  8) ; //  8 = FromHardProcess
+
+	if (theBigTree.genpart_pdg->at(igen) == 23 && isZLast && isZPrompt) // Z0 + isLast + isPrompt
```

Comment:
Thanks Louis! I have a very naive question that originates from my ignorance about MC generation. Should't we require isHardProcess also for the real Z boson?

Reply:
I don't think we should expect any Z boson that is not from the hard process. But it cannot hurt  to double-check that I suppose

Reply:
You're probably right. Thanks!